### PR TITLE
[draft] Rework the way we treat opaques in `check_opaque_meets_bounds`

### PR DESCRIPTION
**Pull request by compiler-errors**
uwu

r? lcnr

### CODE REVIEW

File: compiler/rustc_infer/src/infer/mod.rs [hunk 1]
```diff
@@ -689,23 +692,23 @@ impl<'tcx> InferCtxtBuilder<'tcx> {
...
+            opaque_type_mode: defining_opaque_types,
             considering_regions,
             skip_leak_check,
             intercrate,
             next_trait_solver,
         } = *self;
         InferCtxt {
             tcx,
-            defining_opaque_types,
+            opaque_type_mode: defining_opaque_types,
```
Comment (lcnr) [inline compiler/rustc_infer/src/infer/mod.rs:711]:
```suggestion
            opaque_type_mode,
```
why the rename?

File: compiler/rustc_infer/src/infer/mod.rs [hunk 2]
```diff
@@ -1234,10 +1237,30 @@ impl<'tcx> InferCtxt<'tcx> {
         self.inner.borrow().opaque_type_storage.opaque_types.clone()
     }
 
-    #[inline(always)]
-    pub fn can_define_opaque_ty(&self, id: impl Into<DefId>) -> bool {
-        let Some(id) = id.into().as_local() else { return false };
-        self.defining_opaque_types.contains(&id)
+    pub fn treat_opaque_ty(&self, reveal: Reveal, def_id: DefId) -> TreatOpaque {
```
Comment (lcnr) [inline compiler/rustc_infer/src/infer/mod.rs:1240]: 😊

File: compiler/rustc_infer/src/infer/opaque_types/mod.rs [hunk 3]
```diff
@@ -130,8 +123,18 @@ impl<'tcx> InferCtxt<'tcx> {
                 //     let x = || foo(); // returns the Opaque assoc with `foo`
                 // }
                 // ```
-                if !self.can_define_opaque_ty(def_id) {
-                    return None;
+                match self.treat_opaque_ty(param_env.reveal(), def_id) {
+                    TreatOpaque::Define => {}
+                    TreatOpaque::Reveal | TreatOpaque::Rigid => return None,
```
Comment (lcnr) [inline compiler/rustc_infer/src/infer/opaque_types/mod.rs:128]:
```suggestion
                    // FIXME: We should normalize with `TreatOpaque::Revael`, either
                    // eagerly, allowing us to ICE here, or lazily by revealing the
                    // hidden type here.
                    TreatOpaque::Reveal | TreatOpaque::Rigid => return None,
```

File: compiler/rustc_middle/src/traits/mod.rs [hunk 4]
```diff
@@ -82,6 +82,13 @@ pub enum Reveal {
     All,
 }
 
+pub enum TreatOpaque {
```
Comment (lcnr) [inline compiler/rustc_middle/src/traits/mod.rs:85]: needs comment

File: compiler/rustc_middle/src/ty/closure.rs [hunk 5]
```diff
@@ -194,15 +194,15 @@ impl<'tcx> CapturedPlace<'tcx> {
 
 #[derive(Copy, Clone, Debug, HashStable)]
 pub struct ClosureTypeInfo<'tcx> {
-    user_provided_sig: ty::CanonicalPolyFnSig<'tcx>,
+    user_provided_sig: &'tcx ty::CanonicalPolyFnSig<'tcx>,
```
Comment (lcnr) [inline compiler/rustc_middle/src/ty/closure.rs:197]: why this change?

File: compiler/rustc_type_ir/src/canonical.rs [hunk 6]
```diff
@@ -19,10 +19,31 @@ pub struct Canonical<I: Interner, V> {
...
+    Copy(bound = ""),
+    Clone(bound = ""),
+    Hash(bound = ""),
+    PartialEq(bound = ""),
+    Eq(bound = ""),
+    Debug(bound = "")
+)]
+#[cfg_attr(feature = "nightly", derive(TyEncodable, TyDecodable, HashStable_NoContext))]
+pub enum OpaqueTypeMode<I: Interner> {
+    Define(I::DefiningOpaqueTypes),
```
Comment (lcnr) [inline compiler/rustc_type_ir/src/canonical.rs:37]: the type alias should get renamed to `I::DefIdList` or sth 🤷

File: tests/ui/closures/issue-72408-nested-closures-exponential.rs [hunk 7]
```diff
@@ -1,5 +1,5 @@
 //@ build-pass
-//@ ignore-compare-mode-next-solver (hangs)
+//@ ignore-test (hangs)
```
Comment (lcnr) [inline tests/ui/closures/issue-72408-nested-closures-exponential.rs:2]: 😊 should probably use a FIXME

**Review by lcnr — COMMENTED** [review]
i like it

File: tests/ui/closures/issue-72408-nested-closures-exponential.rs (hunk 7 above)
Comment (compiler-errors) [reply tests/ui/closures/issue-72408-nested-closures-exponential.rs:2]: well this should be fixed before this lands, dont think we can land in this state bc it will definitely break a real crate

File: compiler/rustc_type_ir/src/canonical.rs (hunk 6 above)
Comment (compiler-errors) [reply compiler/rustc_type_ir/src/canonical.rs:37]: Yeah

File: compiler/rustc_middle/src/ty/closure.rs (hunk 5 above)
Comment (compiler-errors) [reply compiler/rustc_middle/src/ty/closure.rs:197]: type sizes get larger so it makes some query machinery complain

File: compiler/rustc_infer/src/infer/mod.rs (hunk 1 above)
Comment (compiler-errors) [reply compiler/rustc_infer/src/infer/mod.rs:711]: i just forgot to defining_opaque_types -> opaque_type_mode
